SignalServiceKit/src/Util/OWSBackgroundTask.m

```diff
@@ -4,14 +4,290 @@
+
+// This property should only be accessed while synchronized on this instance.
+@property (nonatomic) NSMutableDictionary<OWSTaskId, BackgroundTaskExpirationBlock> *expirationMap;
+
+// This property should only be accessed while synchronized on this instance.
+@property (nonatomic) unsigned long long idCounter;
+
+// Note that this flag is set a little early in "will resign active".
+//
+// This property should only be accessed while synchronized on this instance.
+@property (nonatomic) BOOL isAppActive;
+
+// We use this timer to provide continuity and reduce churn,
+// so that if one OWSBackgroundTask ends right before another
+// begins, we use a single uninterrupted background that
+// spans their lifetimes.
+//
+// This property should only be accessed while synchronized on this instance.
+@property (nonatomic, nullable) NSTimer *continuityTimer;
+
+@end
+
+#pragma mark -
+
+@implementation OWSBackgroundTaskManager
+
++ (instancetype)sharedManager
+{
+    static OWSBackgroundTaskManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    OWSAssertIsOnMainThread();
+
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    self.backgroundTaskId = UIBackgroundTaskInvalid;
+    self.expirationMap = [NSMutableDictionary new];
+    self.idCounter = 0;
+    self.isAppActive = CurrentAppContext().isMainAppAndActive;
+
+    OWSSingletonAssert();
+
+    [self observeNotifications];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)observeNotifications
+{
+    if (!CurrentAppContext().isMainApp) {
+        return;
+    }
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidBecomeActive:)
+                                                 name:OWSApplicationDidBecomeActiveNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationWillResignActive:)
+                                                 name:OWSApplicationWillResignActiveNotification
+                                               object:nil];
+}
+
+- (void)applicationDidBecomeActive:(UIApplication *)application
+{
+    OWSAssertIsOnMainThread();
+
+    @synchronized(self)
+    {
+        self.isAppActive = YES;
+
+        [self ensureBackgroundTaskState];
+    }
+}
+
+- (void)applicationWillResignActive:(UIApplication *)application
+{
+    OWSAssertIsOnMainThread();
+
+    @synchronized(self)
+    {
+        self.isAppActive = NO;
+
+        [self ensureBackgroundTaskState];
+    }
+}
+
+// Returns nil if adding this task _should have_ started a
```

Comment:
This method registers a new task with this manager.  We only bother requesting a background task from iOS if the app is inactive (or about to become inactive), so this will often not start a background task.

I've elaborated the comment.